trie-db/src/lib.rs

```diff
@@ -141,37 +142,76 @@ where
-///
-/// This is implemented for any &mut recorder (where the query will return
-/// a DBValue), any function taking raw bytes (where no recording will be made),
-/// or any tuple of (&mut Recorder, FnOnce(&[u8]))
 pub trait Query<H: Hasher> {
 	/// Output item.
 	type Item;
 
 	/// Decode a byte-slice into the desired item.
 	fn decode(self, data: &[u8]) -> Self::Item;
+}
 
-	/// Record that a node has been passed through.
-	fn record(&mut self, _hash: &H::Out, _data: &[u8], _depth: u32) {}
+/// Used to report the trie access to the [`TrieRecorder`].
+///
+/// As the trie can use a [`TrieCache`], there are multiple kinds of accesses.
+/// If a cache is used, [`Self::Key`] and [`Self::NodeOwned`] are possible
+/// values. Otherwise only [`Self::EncodedNode`] is a possible value.
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum TrieAccess<'a, H> {
+	/// The given [`NodeOwned`] was accessed using its `hash`.
+	NodeOwned { hash: H, node_owned: &'a NodeOwned<H> },
+	/// The given `encoded_node` was accessed using its `hash`.
+	EncodedNode { hash: H, encoded_node: rstd::borrow::Cow<'a, [u8]> },
+	/// The given `value` was accessed using its `hash`.
+	///
+	/// The given `full_key` is the key to access this value in the trie.
+	///
+	/// Should map to [`RecordedForKey::Value`] when checking the recorder.
+	Value { hash: H, value: rstd::borrow::Cow<'a, [u8]>, full_key: &'a [u8] },
+	/// The hash of the value for the given `full_key` was accessed.
+	///
+	/// Should map to [`RecordedForKey::Hash`] when checking the recorder.
+	Hash { full_key: &'a [u8] },
+	/// The value/hash for `full_key` was accessed, but it couldn't be found in the trie.
+	///
+	/// Should map to [`RecordedForKey::Value`] when checking the recorder.
+	NonExisting { full_key: &'a [u8] },
 }
 
-impl<'a, H: Hasher> Query<H> for &'a mut Recorder<H::Out> {
-	type Item = DBValue;
-	fn decode(self, value: &[u8]) -> DBValue {
-		value.to_vec()
-	}
-	fn record(&mut self, hash: &H::Out, data: &[u8], depth: u32) {
-		(&mut **self).record(hash, data, depth);
+/// Result of [`TrieRecorder::trie_nodes_recorded_for_key`].
+#[derive(Debug, Clone, Copy)]
+pub enum RecordedForKey {
+	/// We recorded all trie nodes up to the value for a storage key.
+	Value,
+	/// We recorded all trie nodes up to the value hash for a storage key.
+	Hash,
+	/// We didn't yet recorded any trie nodes for a storage key.
+	Nothing,
+}
+
+impl RecordedForKey {
+	/// Is `self` equal to [`Self::Nothing`]?
+	pub fn is_nothing(&self) -> bool {
+		matches!(self, Self::Nothing)
 	}
 }
 
+/// A trie recorder that can be used to record all kind of trie accesses.
+pub trait TrieRecorder<H> {
+	/// Record the given [`TrieAccess`].
+	fn record<'a>(&mut self, access: TrieAccess<'a, H>);
+
+	/// Check if we have recorded any trie nodes for the given `key`.
+	///
+	/// Returns [`RecordedForKey`] to express the state of the recorded trie nodes.
+	fn trie_nodes_recorded_for_key(&self, key: &[u8]) -> RecordedForKey;
```

Comment:
Makes the trait closer to a cache, so I was wondering a bt if recorder and cache should simply be merged, but I really don't  know.

Reply:
I mean it not really makes it closer to the cache. It just gives the user an opportunity to check the status of a certain key to then be able to use the cache if allowed.